flink-python/src/main/java/org/apache/flink/table/runtime/operators/python/BaseRowPythonTableFunctionOperator.java

```diff
@@ -0,0 +1,148 @@
+import org.apache.beam.sdk.fn.data.FnDataReceiver;
+
+import java.io.IOException;
+
+/**
+ * The Python {@link TableFunction} operator for the blink planner.
+ */
+@Internal
+public class BaseRowPythonTableFunctionOperator
+	extends AbstractPythonTableFunctionOperator<BaseRow, BaseRow, BaseRow> {
+
+
+	private static final long serialVersionUID = 1L;
+
+	/**
+	 * The collector used to collect records.
+	 */
+	private transient StreamRecordBaseRowWrappingCollector baseRowWrapper;
+
+	/**
+	 * The JoinedRow reused holding the execution result.
+	 */
+	private transient JoinedRow reuseJoinedRow;
+
+	/**
+	 * The Projection which projects the udtf input fields from the input row.
+	 */
+	private transient Projection<BaseRow, BinaryRow> udtfInputProjection;
+
+	/**
+	 * The TypeSerializer for udtf execution results.
+	 */
+	private transient TypeSerializer<BaseRow> udtfOutputTypeSerializer;
+
+	public BaseRowPythonTableFunctionOperator(
+		Configuration config,
+		PythonFunctionInfo tableFunction,
+		RowType inputType,
+		RowType outputType,
+		int[] udtfInputOffsets) {
+		super(config, tableFunction, inputType, outputType, udtfInputOffsets);
+	}
+
+	@Override
+	@SuppressWarnings("unchecked")
+	public void open() throws Exception {
+		super.open();
+		baseRowWrapper = new StreamRecordBaseRowWrappingCollector(output);
+		reuseJoinedRow = new JoinedRow();
+
+		udtfInputProjection = createUdtfInputProjection();
+		udtfOutputTypeSerializer = PythonTypeUtils.toBlinkTypeSerializer(userDefinedFunctionOutputType);
+	}
+
+	@Override
+	public void bufferInput(BaseRow input) {
+		forwardedInputQueue.add(input);
+	}
+
+	@Override
+	public BaseRow getUdfInput(BaseRow element) {
+		return udtfInputProjection.apply(element);
+	}
+
+	@Override
+	public PythonFunctionRunner<BaseRow> createPythonFunctionRunner(
+		FnDataReceiver<byte[]> resultReceiver,
+		PythonEnvironmentManager pythonEnvironmentManager) {
+		return new BaseRowPythonTableFunctionRunner(
+			getRuntimeContext().getTaskName(),
+			resultReceiver,
+			tableFunction,
+			pythonEnvironmentManager,
+			userDefinedFunctionInputType,
+			userDefinedFunctionOutputType);
+	}
+
+	private Projection<BaseRow, BinaryRow> createUdtfInputProjection() {
+		final GeneratedProjection generatedProjection = ProjectionCodeGenerator.generateProjection(
+			CodeGeneratorContext.apply(new TableConfig()),
+			"UdtfInputProjection",
+			inputType,
+			userDefinedFunctionInputType,
+			userDefinedFunctionInputOffsets);
+		// noinspection unchecked
+		return generatedProjection.newInstance(Thread.currentThread().getContextClassLoader());
+	}
+
+	@Override
+	public void emitResults() throws IOException {
+		BaseRow input = null;
+		byte[] rawUdtfResult;
+		while ((rawUdtfResult = userDefinedFunctionResultQueue.poll()) != null) {
+			if (input == null) {
+				input = forwardedInputQueue.poll();
+			}
+			boolean isFinishResult = isFinishResult(rawUdtfResult);
+			if (isFinishResult) {
+				input = forwardedInputQueue.poll();
+			}
```

Comment:
The logic here contains a bug. Take the following case as an example:
- There comes the first record. 
- However, the result is empty, so the Python side only sends a finished result. 

In this case, if it is a left join, we should output the result(input append null). 
Maybe it's better to also add some test for this.

Reply:
Good Catch!